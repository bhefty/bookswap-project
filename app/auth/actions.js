import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGOUT,
  SET_SESSION
} from './constants'

export const loginRequest = () => ({
  type: LOGIN_REQUEST
})

export const loginSuccess = () => ({
  type: LOGIN_SUCCESS
})

export const loginFailure = (error) => ({
  type: LOGIN_FAILURE,
  error
})

export const logout = () => ({
  type: LOGOUT
})

export const setSession = (authResult) => ({
  type: SET_SESSION,
  authResult
})
