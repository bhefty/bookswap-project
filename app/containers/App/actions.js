/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_TRANSLATION,
  LOAD_TRANSLATION_SUCCESS,
  LOAD_TRANSLATION_ERROR
} from './constants'

/**
 * Load the translation, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_TRANSLATION
 */
export function loadTranslation () {
  return {
    type: LOAD_TRANSLATION
  }
}

/**
 * Dispatched when the translation is loaded by the request saga
 *
 * @param  {array} translation The translation data
 * @param  {string} text The current text
 *
 * @return {object} An action object with a type of LOAD_TRANSLATION_SUCCESS passing the translation
 */
export function translationLoaded (translation, text) {
  return {
    type: LOAD_TRANSLATION_SUCCESS,
    translation,
    text
  }
}

/**
 * Dispatched when loading the translation fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_TRANSLATION_ERROR passing the error
 */
export function translationLoadingError (error) {
  return {
    type: LOAD_TRANSLATION_ERROR,
    error
  }
}
