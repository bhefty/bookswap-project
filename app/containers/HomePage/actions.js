import * as types from './constants'

export const changeText = (text) => ({
  type: types.CHANGE_TEXT,
  text
})

export const increment = () => ({
  type: types.INCREMENT
})

export const decrement = () => ({
  type: types.DECREMENT
})
