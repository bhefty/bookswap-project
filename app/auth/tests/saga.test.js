import sagaHelper from 'redux-saga-testing'
import { call, put, take } from 'redux-saga/effects'
import { push } from 'react-router-redux'
import localStorage from 'localStorage'

import {
  loginRequestSaga,
  logoutSaga,
  watchLoginRequest,
  watchLoginSuccess,
  watchLoginFailure,
  watchLogout,
  watchSetSession,
  setSessionSaga
} from '../saga'

import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGOUT,
  SET_SESSION
} from '../constants'

import { loginSuccess, loginFailure } from '../actions'

const auth = {
  login: jest.fn(),
  logout: jest.fn()
}

const mockApiData = {
  access_token: '123abc',
  id_token: '456def',
  expiresIn: 1234567890
}

describe('authSagas', () => {
  describe('loginRequestSaga', () => {
    const it = sagaHelper(loginRequestSaga())

    it('should call auth.login', result => {
      const resultJSON = JSON.stringify(result)
      expect(resultJSON).toEqual(JSON.stringify(call(auth.login)))
    })

    it('should do nothing', result => {
      expect(result).toBeUndefined()
    })
  })

  describe('loginRequestSaga - Error handling', () => {
    const it = sagaHelper(loginRequestSaga())

    it('should call auth.login', result => {
      const resultJSON = JSON.stringify(result)
      expect(resultJSON).toEqual(JSON.stringify(call(auth.login)))
      return new Error('Something went wrong')
    })

    it('should trigger loginFailure action with the error message', result => {
      expect(result).toEqual(put(loginFailure(new Error('Something went wrong'))))
    })

    it('should do nothing', result => {
      expect(result).toBeUndefined()
    })
  })

  describe('setSessionSaga', () => {
    const generator = setSessionSaga()
    const selectDescriptor = generator.next().value

    it('should select authResult from state', () => {
      expect(selectDescriptor).toMatchSnapshot()
    })

    const putDescriptor = generator.next(mockApiData).value

    it('should trigger loginSuccess', () => {
      expect(putDescriptor).toEqual(put(loginSuccess()))
    })

    const endDescriptor = generator.next().value

    it('should do nothing', () => {
      expect(endDescriptor).toBeUndefined()
    })
  })

  describe('logoutSaga', () => {
    it('should remove tokens from local storage', () => {
      localStorage.setItem('access_token', 'abc123')
      localStorage.setItem('id_token', '123abc')
      localStorage.setItem('expires_at', '7200')

      expect(localStorage.getItem('access_token')).toEqual('abc123')
      expect(localStorage.getItem('id_token')).toEqual('123abc')
      expect(localStorage.getItem('expires_at')).toEqual('7200')

      const generator = logoutSaga()
      generator.next()

      expect(localStorage.getItem('access_token')).toEqual(null)
      expect(localStorage.getItem('id_token')).toEqual(null)
      expect(localStorage.getItem('expires_at')).toEqual(null)
    })
  })

  describe('watchLoginRequest', () => {
    const it = sagaHelper(watchLoginRequest())

    it('should take LOGIN_REQUEST', result => {
      expect(result).toEqual(take(LOGIN_REQUEST))
    })

    it('should call loginRequestSaga', result => {
      expect(result).toEqual(call(loginRequestSaga))
    })

    it('should do nothing', result => {
      expect(result).toBeUndefined()
    })
  })

  describe('watchLoginSuccess', () => {
    const it = sagaHelper(watchLoginSuccess())

    it('should take LOGIN_SUCCESS', result => {
      expect(result).toEqual(take(LOGIN_SUCCESS))
    })

    it('should trigger redirect to \'/\'', result => {
      expect(result).toEqual(put(push('/dashboard')))
    })

    it('should do nothing', result => {
      expect(result).toBeUndefined()
    })
  })

  describe('watchLoginFailure', () => {
    const it = sagaHelper(watchLoginFailure())

    it('should take LOGIN_FAILURE', result => {
      expect(result).toEqual(take(LOGIN_FAILURE))
    })

    it('should trigger redirect to \'/\'', result => {
      expect(result).toEqual(put(push('/')))
    })

    it('should do nothing', result => {
      expect(result).toBeUndefined()
    })
  })

  describe('watchLogout', () => {
    const it = sagaHelper(watchLogout())

    it('should take LOGOUT', result => {
      expect(result).toEqual(take(LOGOUT))
    })

    it('should call auth.logout', result => {
      const resultJSON = JSON.stringify(result)
      expect(resultJSON).toEqual(JSON.stringify(call(logoutSaga)))
    })

    it('should trigger redirect to \'/\'', result => {
      expect(result).toEqual(put(push('/')))
    })

    it('should do nothing', result => {
      expect(result).toBeUndefined()
    })
  })

  describe('watchSetSession', () => {
    const it = sagaHelper(watchSetSession())

    it('should take SET_SESSION', result => {
      expect(result).toEqual(take(SET_SESSION))
    })

    it('should call setSessionSaga', result => {
      expect(result).toEqual(call(setSessionSaga))
    })

    it('should do nothing', result => {
      expect(result).toBeUndefined()
    })
  })
})
