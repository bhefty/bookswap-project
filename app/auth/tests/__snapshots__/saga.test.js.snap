// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authSagas setSessionSaga should select authResult from state 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;
